### ISSUE TITLE
Regression with using high DPI images in wxImageList

### ISSUE THREAD

**Issue by vadz** [wxWidgets/wxWidgets#22189]
Quoting [this post](https://groups.google.com/g/wx-dev/c/2FKrdBguK-w/m/IFjjF7GxHQAJ):

> The root cause of the bitmaps not showing at all is due to a call to:
>
> wxArtProvider::GetBitmap(wxART_MAKE_ART_ID_FROM_STR("wxART_NORMAL_FILE"), wxART_LIST, wxDefaultSize);
>
> returning a wxBitmap of size 16x16 (scale 1) instead of 24x24 (scale 1) which it did previously.  We detect when the art provider doesn't provide the needed 16x16 and we scaled this to 32x32 with scale factor 2 (on Retina displays) to match all the other images we were adding to the wxImageList.    With it returning 16x16, we were not scaling and setting it into the wxImageList which records the scalefactor of the first bitmap added and ONLY allows bitmaps of the same scaling factor to be added.  I'm not sure if that restriction is still valid.  (or if wxImageList should use wxBitmapBundles instead).     Anyway, if I always make sure the bitmap from the artprovider is the 32x32x2.0 size, the images appear.   HOWEVER, the images are fuzzy.  It looks like it's taking the 32x32x2 icons and resizing them down to the exact 16x16x1 before drawing which on OSX's retina display loses the HiDPI.  

Not clear if this is Mac-specific or also affects wxGTK3 (wxMSW uses a different `wxImageList` implementation).

**Comment by dkulp** [comment]
Did a bisect.   The first bad commit is 4073f3b9bf885d347bf10b8174e854be3f66c7ab

**Comment by vadz** [comment]
Thanks, but this is really strange, this commit only changes wxDVC:
```
% git log -1 --format=%s --stat 4073f3b9bf885d347bf10b8174e854be3f66c7ab
Use wxBitmapBundle in wxDataViewCtrl-related classes

 include/wx/dataview.h         | 69 +++++++++++++++++++++++++++++++++++++++++++++++----------------------
 include/wx/dvrenderers.h      | 19 ++++++++++++-------
 interface/wx/dataview.h       | 63 ++++++++++++++++++++++++++++++++++++++++++++-------------------
 samples/dataview/dataview.cpp | 14 ++++++--------
 samples/dataview/mymodels.cpp | 25 ++++---------------------
 samples/dataview/mymodels.h   | 11 ++++++++++-
 samples/dataview/null.xpm     | 30 ------------------------------
 samples/dataview/wx_small.xpm | 28 ----------------------------
 src/common/datavcmn.cpp       | 68 ++++++++++++++++++++++++++++++++++++++------------------------------
 src/generic/datavgen.cpp      | 20 ++++++++++++--------
 10 files changed, 173 insertions(+), 174 deletions(-)
```

How could it possibly affect this?

Anyhow, I think there are 3 different problems here:

1. Change of the size of bitmap returned by `wxArtProvider`.
2. Fuzzy drawing of images with scale factor 2.
3. Silently dropping images with different scale factors in `wxImageList`.

I think (3) was always there and (2) might not be new neither, but (1) must have changed recently. It really can't have changed as the result of this commit, however.

I'll try to look at this a.s.a.p. but not today, sorry.

**Comment by dkulp** [comment]
I'm trying to track down #2.    #3 is not new but exposed by #1, but #1 is easily worked around so I'm trying to figure out #2.  The commit prior to 4073f3b9bf885d347bf10b8174e854be3f66c7ab ( 90bbde87773337ba471fa584df3aee72e43364ee ) draws the images in the list nice and sharp as they should be.

**Comment by dkulp** [comment]
To provide more details....     I can confirm that the wxIcon that is set into the cell to draw on OSX is a 16x16x1 image instead of the original 32x32x2.     In wxDataViewIconTextRenderer::MacRender() (line 3216 of dataview.mm), I can see that the wxBitmapBundle that the iconText is holding does have the 32x32x2 bitmap.  Thus, it's OK at that point.   However, it's calling the GetIcon() on the iconText which calls "m_bitmap.GetIcon(wxDefaultSize);".   That is resulting in a call to " wxBitmapBundleImplSet::GetBitmap(const wxSize& size)" (line 220 of src/osx/core/bmpbndl.mm) with size 16x16 as that is the "m_sizeDefault" for the BitmapBundle.

There is only a single entry in the bundle which is 32x32x2.  The call to entry.bitmap.GetSize returns 32x32.  This size is LARGER than the passed in size of 16x16 (note that scale is NOT consulted in GetBitmap at all) and thus a new entry is created (line 249) with the size of 16x16x1.   This is then returned.   

I hope that helps.

**Comment by dkulp** [comment]
One more follow up, if I change the call to 'const wxSize sizeThis = entry.bitmap.GetSize();' to 'const wxSize sizeThis = entry.bitmap.GetDIPSize();' then everything draws correctly.  But I'm not sure if the size passed into wxBitmapBundleImplSet::GetBitmap is supposed to be a physical size or logical size.

**Comment by vadz** [comment]
All these issues should be fixed by #22211. @dkulp please let me know if you still see any problems with it and, if you do, how can I reproduce them. TIA!
